source/slang/hlsl.meta.slang

```diff
@@ -6885,3 +6894,115 @@ uint3 cudaBlockIdx();
 __target_intrinsic(cuda, "(blockDim)")
 [__readNone]
 uint3 cudaBlockDim();
+
+//
+// Workgroup cooperation
+//
+
+//
+// `saturated_cooperation(c, f, s, u)` will call `f(s, u)` if not all lanes in the
+// workgroup are currently executing. however if all lanes are saturated, then
+// for each unique `s` across all the active lanes `c(s, u)` is called. The
+// return value is the one corresponding to the input `s` from this lane.
+//
+// Adjacent calls to saturated_cooperation are subject to fusion, i.e.
+//      saturated_cooperation(c1, f1, s, u1);
+//      saturated_cooperation(c2, f2, s, u2);
+// will be transformed to:
+//      saturated_cooperation(c1c2, f1f2, s, u1u2);
+// where
+//      c1c2 is a function which calls c1(s, u1) and then c2(s, u2);
+//      f1f2 is a function which calls f1(s, u1) and then f2(s, u2);
+//
+// When the input differs, calls are fused
+//      saturated_cooperation(c1, f1, s1, u1);
+//      saturated_cooperation(c2, f2, s2, u2);
+// will be transformed to:
+//      saturated_cooperation(c1c2, f1f2, s1s2, u1u2);
+// where
+//      s1s2 is a tuple of s1 and s2
+//      c1c2 is a function which calls c1(s1, u1) and then c2(s2, u2);
+//      f1f2 is a function which calls f1(s1, u1) and then f2(s2, u2);
+// Note that in this case, we will make a call to c1c2 for every unique pair
+// s1s2 across all lanes
+//
+// (This fusion takes place in the fuse-satcoop pass, and as such any changes to
+// the signature or behavior of this function should be adjusted for there).
+//
+func saturated_cooperation<A : __BuiltinType, B, C>(
+    cooperate : functype (A, B) -> C,
+    fallback : functype (A, B) -> C,
+    A input,
+    B otherArg)
+    -> C
+{
+    return saturated_cooperation_using(cooperate, fallback, __WaveMatchBuitin<A>, __WaveReadLaneAtBuiltin<A>, input, otherArg);
+}
+
+// These two functions are a temporary (circa May 2023) workaround to the fact
+// that we can't deduce which overload to pass to saturated_cooperation_using
+// in the call above
+[__unsafeForceInlineEarly]
+func __WaveMatchBuitin<T : __BuiltinType>(T t) -> uint4
+{
+    return WaveMatch(t);
+}
+[__unsafeForceInlineEarly]
+func __WaveReadLaneAtBuiltin<T : __BuiltinType>(T t, int i) -> T
+{
+    return WaveReadLaneAt(t, i);
+}
+
+//
+// saturated_cooperation, but you're able to specify manually the functions:
+//
+// waveMatch: a function to return a mask of lanes with the same input as this one
+// broadcast: a function which returns the value passed into it on the specified lane
+//
+func saturated_cooperation_using<A, B, C>(
+    cooperate : functype (A, B) -> C,
+    fallback : functype (A, B) -> C,
+    waveMatch : functype (A) -> uint4,
+    broadcast : functype (A, int) -> A,
+    A input,
+    B otherArg)
```

Comment:
What if we have more than one B? For example, matrix multiplication will have two arguments instead of 1.
We need to provide overloads of this function with different numbers of arguments.

Reply:
I *think* karthik said that a single parameter would be ok for their use case. Either way, shouldn't be hard to add an overload I guess, although I would prefer if people wrapped their multiple parameters in a struct (or eventually, anonymous tuple).